### PR TITLE
Use builtins for byteswapping

### PR DESCRIPTION
**Pull request by nimish**
Make use of XCHG/BSWAP on x86, REV16 and REV on ARMv6-m and above, and whatever other optimized instructions on other platforms.

## Summary
Use compiler builtins where available
## Impact
A little speed boost

### CODE REVIEW

File: include/netinet/in.h [hunk 1]
```diff
@@ -217,26 +217,30 @@
...
- *   Copyright (c) 2001-2003, Adam Dunkels.
- *   All rights reserved.
  */
 
 #ifdef CONFIG_ENDIAN_BIG
 # define HTONS(ns) (ns)
 # define HTONL(nl) (nl)
 #else
+# ifdef __has_builtin && __has_builtin(__builtin_bswap16)
+#  define HTONS(ns) ((unsigned short)__builtin_bswap16(((unsigned short) (ns))))
```
Comment (xiaoxiang781216) [inline include/netinet/in.h:227]: it's better to update the related macros in endian.h and forward HTONS to them.

File: include/netinet/in.h [hunk 2]
```diff
@@ -232,7 +232,7 @@
      ((((unsigned short)(ns)) >> 8) & 0x00ff))
 # endif
 # ifdef __has_builtin && __has_builtin(__builtin_bswap32)
-#  define HTONL(nl) ((unsigned long)__builtin_bswap32(((unsigned short) (nl))))
+#  define HTONL(nl) ((unsigned long)__builtin_bswap32(((unsigned long) (nl))))
```
Comment (xiaoxiang781216) [inline include/netinet/in.h:235]: merge to previous patch

File: include/netinet/in.h (hunk 1 above)
Comment (pkarashchenko) [reply include/netinet/in.h:227]: This is purely compiler dependent, so better to move compiler stuff to `compiler..h`

Comment (pkarashchenko) [reply include/netinet/in.h:227]: Maybe adding something similar `CONFIG_HAVE_BUILTIN_CTZ` may be better.

File: include/netinet/in.h (hunk 2 above)
Comment (nimish) [reply include/netinet/in.h]: squash merge it at the end?

File: include/netinet/in.h (hunk 1 above)
Comment (nimish) [reply include/netinet/in.h]: Done, moved to other headers

File: include/netinet/in.h (hunk 2 above)
Comment (nimish) [reply include/netinet/in.h]: done.

File: include/nuttx/compiler.h [hunk 3]
```diff
@@ -69,6 +69,36 @@
...
+# if  __has_builtin(__builtin_ctz)
+#  define CONFIG_HAVE_BUILTIN_CTZ 1
+# endif
+# if  __has_builtin(__builtin_clz)
+#  define CONFIG_HAVE_BUILTIN_CLZ 1
+# endif
+# if  __has_builtin(__builtin_popcount)
+#  define CONFIG_HAVE_BUILTIN_POPCOUNT 1
+# endif
+#endif
```
Comment (pkarashchenko) [inline include/nuttx/compiler.h:99]: Please move all this block under `#ifdef __GNUC__`.
Also according https://stackoverflow.com/questions/54079257/how-to-check-builtin-function-is-available-on-gcc the `__has_builtin` is a new GCC feature while the old version might still have builtin functions available, but will fail `__has_builtin` check. It is better to rely on the GCC version.

File: include/netinet/in.h [hunk 4]
```diff
@@ -217,26 +217,16 @@
...
- *   Copyright (c) 2001-2003, Adam Dunkels.
- *   All rights reserved.
  */
 
 #ifdef CONFIG_ENDIAN_BIG
 # define HTONS(ns) (ns)
 # define HTONL(nl) (nl)
+#elif defined (CONFIG_ENDIAN_LITTLE)
+# define HTONS(ns) ((uint16_t)__swap_uint16(((uint16_t) (ns))))
+# define HTONL(nl) ((uint32_t)__swap_uint32(((uint32_t) (nl))))
```
Comment (pkarashchenko) [inline include/netinet/in.h:227]:
```suggestion
# define HTONS(ns) (__swap_uint16(ns))
# define HTONL(nl) (__swap_uint32(nl))
```

File: include/netinet/in.h [hunk 5]
```diff
@@ -217,26 +217,16 @@
...
- *
- *   Author Adam Dunkels <adam@dunkels.com>
- *   Copyright (c) 2001-2003, Adam Dunkels.
- *   All rights reserved.
  */
 
 #ifdef CONFIG_ENDIAN_BIG
 # define HTONS(ns) (ns)
 # define HTONL(nl) (nl)
+#elif defined (CONFIG_ENDIAN_LITTLE)
```
Comment (pkarashchenko) [inline include/netinet/in.h:225]:
```suggestion
#elif defined(CONFIG_ENDIAN_LITTLE)
```

File: include/netinet/in.h [hunk 6]
```diff
@@ -217,26 +217,16 @@
...
-  (unsigned short) \
-    (((((unsigned short)(ns)) & 0x00ff) << 8) | \
-     ((((unsigned short)(ns)) >> 8) & 0x00ff))
-# define HTONL(nl) \
-  (unsigned long) \
-    (((((unsigned long)(nl)) & 0x000000ffUL) << 24) | \
-     ((((unsigned long)(nl)) & 0x0000ff00UL) <<  8) | \
-     ((((unsigned long)(nl)) & 0x00ff0000UL) >>  8) | \
-     ((((unsigned long)(nl)) & 0xff000000UL) >> 24))
+# error "PDP-11s are not supported."
```
Comment (pkarashchenko) [inline include/netinet/in.h:229]:
```suggestion
#  error "PDP-11s are not supported."
```

File: include/netinet/in.h [hunk 7]
```diff
@@ -217,26 +217,16 @@
...
  * for uIP:
- *
- *   Author Adam Dunkels <adam@dunkels.com>
- *   Copyright (c) 2001-2003, Adam Dunkels.
- *   All rights reserved.
  */
 
 #ifdef CONFIG_ENDIAN_BIG
 # define HTONS(ns) (ns)
 # define HTONL(nl) (nl)
```
Comment (pkarashchenko) [inline include/netinet/in.h:224]:
```suggestion
#  define HTONS(ns) (ns)
#  define HTONL(nl) (nl)
```

File: include/netinet/in.h [hunk 8]
```diff
@@ -217,26 +217,16 @@
 /* This macro to convert a 16/32-bit constant values quantity from host byte
  * order to network byte order.  The 16-bit version of this macro is required
  * for uIP:
- *
- *   Author Adam Dunkels <adam@dunkels.com>
- *   Copyright (c) 2001-2003, Adam Dunkels.
- *   All rights reserved.
```
Comment (pkarashchenko) [inline include/netinet/in.h:223]: Why this is removed?

File: include/endian.h [hunk 9]
```diff
@@ -52,21 +52,50 @@
 /* Common byte swapping macros */
 
 #define __SWAP_UINT16_ISMACRO 1
-#undef  __SWAP_UINT32_ISMACRO
+#define __SWAP_UINT32_ISMACRO 1
+#define __SWAP_UINT64_ISMACRO 1
```
Comment (xiaoxiang781216) [inline include/endian.h:56]: let's remove all __SWAP_UINTXX_ISMACRO and libs/libc/endian/lib_swapxx.c? it doesn't make sense to keep them if we always enable the macro version.

File: include/netinet/in.h (hunk 8 above)
Comment (nimish) [reply include/netinet/in.h:223]: * Unclear if we need it given we aren't using the original swaps
* It appears to conflict with the top-level copyright header (all rights reserved vs Apache)

No objection to keeping it.

File: include/nuttx/compiler.h (hunk 3 above)
Comment (nimish) [reply include/nuttx/compiler.h:99]: I just switched these to checking the symbol itself, since `__GNUC__` can be disabled on clang but the builtins might still exist.

**Comment by xiaoxiang781216** [comment]
@nimish please squash the change into one patch:
git rebase --interactive HEAD~9
and chang pick to sqaush except the first one

File: include/endian.h [hunk 10]
```diff
@@ -52,21 +52,50 @@
...
+
 
 #ifdef __SWAP_UINT16_ISMACRO
-#  define __swap_uint16(n) \
+#  ifdef CONFIG_HAVE_BUILTIN_BSWAP16
+#    define __swap_uint16(n) ((uint16_t)__builtin_bswap16((uint32_t)(n)))
```
Comment (pkarashchenko) [inline include/endian.h:61]:
```suggestion
#    define __swap_uint16 __builtin_bswap16
```

File: include/endian.h [hunk 11]
```diff
@@ -52,21 +52,50 @@
...
+#    define __swap_uint16(n) \
     (uint16_t)(((((uint16_t)(n)) & 0x00ff) << 8) | \
                ((((uint16_t)(n)) >> 8) & 0x00ff))
+#  endif
 #endif
 
 #ifdef __SWAP_UINT32_ISMACRO
+#ifdef CONFIG_HAVE_BUILTIN_BSWAP32
+#  define __swap_uint32(n) \
+    ((uint32_t)__builtin_bswap32((uint32_t)(n)))
```
Comment (pkarashchenko) [inline include/endian.h:72]:
```suggestion
#  define __swap_uint32 __builtin_bswap32
```

File: include/endian.h [hunk 12]
```diff
@@ -52,21 +52,50 @@
...
                ((((uint32_t)(n)) & 0x0000ff00UL) <<  8) | \
                ((((uint32_t)(n)) & 0x00ff0000UL) >>  8) | \
                ((((uint32_t)(n)) & 0xff000000UL) >> 24))
 #endif
+#endif
+
+#ifdef __SWAP_UINT64_ISMACRO && CONFIG_HAVE_LONG_LONG
+#ifdef CONFIG_HAVE_BUILTIN_BSWAP64
+#  define __swap_uint64(n) \
+    ((uint64_t)__builtin_bswap64((uint64_t)(n)))
```
Comment (pkarashchenko) [inline include/endian.h:85]:
```suggestion
#  define __swap_uint64 __builtin_bswap64
```

File: include/nuttx/compiler.h (hunk 3 above)
Comment (pkarashchenko) [reply include/nuttx/compiler.h:99]: I think Clag also defines `__GNUC__`

File: libs/libc/string/lib_ffsll.c [hunk 13]
```diff
@@ -57,10 +57,12 @@ int ffsll(long long j)
 
   if (j != 0)
     {
-#ifdef CONFIG_HAVE_BUILTIN_CTZ
+#ifdef CONFIG_HAVE_BUILTIN_FFSLL
+      ret = __builtin_ffs(j);
```
Comment (pkarashchenko) [inline libs/libc/string/lib_ffsll.c:61]: Should it be

```
      ret = __builtin_ffsll(j);
```
?

File: libs/libc/string/lib_ffsl.c [hunk 14]
```diff
@@ -48,14 +48,15 @@
...
-
 int ffsl(long j)
 {
   int ret = 0;
 
   if (j != 0)
     {
-#ifdef CONFIG_HAVE_BUILTIN_CTZ
+#ifdef CONFIG_HAVE_BUILTIN_FFSL
+      ret = __builtin_ffs(j);
```
Comment (pkarashchenko) [inline libs/libc/string/lib_ffsl.c:58]: Should it be

```
      ret = __builtin_ffsl(j);
```
?

File: include/nuttx/compiler.h [hunk 15]
```diff
@@ -61,6 +61,14 @@
 #  define CONFIG_C99_BOOL 1
 #endif
 
+#if defined(__STDC_VERSION__) && __STDC_VERSION__ >= 201112L
+#  define CONFIG_C11_BOOL 1
+#endif
+
+#if defined(__STDC_VERSION__) && __STDC_VERSION__ >= 201710L
+#  define CONFIG_C17_BOOL 1
```
Comment (pkarashchenko) [inline include/nuttx/compiler.h:69]: Where is it used?

File: include/nuttx/compiler.h [hunk 16]
```diff
@@ -61,6 +61,14 @@
 #  define CONFIG_C99_BOOL 1
 #endif
 
+#if defined(__STDC_VERSION__) && __STDC_VERSION__ >= 201112L
+#  define CONFIG_C11_BOOL 1
```
Comment (pkarashchenko) [inline include/nuttx/compiler.h:65]: Where is it used?

File: include/netinet/in.h [hunk 17]
```diff
@@ -217,26 +217,14 @@
...
-    (((((unsigned short)(ns)) & 0x00ff) << 8) | \
-     ((((unsigned short)(ns)) >> 8) & 0x00ff))
-# define HTONL(nl) \
-  (unsigned long) \
-    (((((unsigned long)(nl)) & 0x000000ffUL) << 24) | \
-     ((((unsigned long)(nl)) & 0x0000ff00UL) <<  8) | \
-     ((((unsigned long)(nl)) & 0x00ff0000UL) >>  8) | \
-     ((((unsigned long)(nl)) & 0xff000000UL) >> 24))
+#  define HTONS(ns) (__swap_uint16(ns))
+#  define HTONL(nl) (__swap_uint32(nl))
```
Comment (pkarashchenko) [inline include/netinet/in.h:227]:
```suggestion
#  define HTONS(ns) __swap_uint16(ns)
#  define HTONL(nl) __swap_uint32(nl)
```
Or
```suggestion
#  define HTONS __swap_uint16
#  define HTONL __swap_uint32
```

File: include/nuttx/compiler.h (hunk 15 above)
Comment (nimish) [reply include/nuttx/compiler.h:69]: This is for future compatibility, in conjunction with the other ones

File: libs/libc/string/lib_ffsll.c (hunk 13 above)
Comment (nimish) [reply libs/libc/string/lib_ffsll.c]: yep

File: libs/libc/string/lib_ffsl.c (hunk 14 above)
Comment (nimish) [reply libs/libc/string/lib_ffsl.c]: yep

File: include/nuttx/compiler.h (hunk 3 above)
Comment (nimish) [reply include/nuttx/compiler.h]: done. Clang does unless you disable it and it may not keep up with GCC's one

File: include/endian.h [hunk 18]
```diff
@@ -51,23 +51,41 @@
...
+#else
 #  define __swap_uint32(n) \
     (uint32_t)(((((uint32_t)(n)) & 0x000000ffUL) << 24) | \
                ((((uint32_t)(n)) & 0x0000ff00UL) <<  8) | \
                ((((uint32_t)(n)) & 0x00ff0000UL) >>  8) | \
                ((((uint32_t)(n)) & 0xff000000UL) >> 24))
 #endif
 
+#ifdef CONFIG_HAVE_LONG_LONG
+#ifdef CONFIG_HAVE_BUILTIN_BSWAP64
```
Comment (xiaoxiang781216) [inline include/endian.h:73]: let's indent the inner #ifdef/#else/#endif by two spaces

File: include/netinet/in.h [hunk 19]
```diff
@@ -217,26 +217,14 @@
...
-  (unsigned short) \
-    (((((unsigned short)(ns)) & 0x00ff) << 8) | \
-     ((((unsigned short)(ns)) >> 8) & 0x00ff))
-# define HTONL(nl) \
-  (unsigned long) \
-    (((((unsigned long)(nl)) & 0x000000ffUL) << 24) | \
-     ((((unsigned long)(nl)) & 0x0000ff00UL) <<  8) | \
-     ((((unsigned long)(nl)) & 0x00ff0000UL) >>  8) | \
-     ((((unsigned long)(nl)) & 0xff000000UL) >> 24))
+#  define HTONS __swap_uint16
```
Comment (xiaoxiang781216) [inline include/netinet/in.h:226]: need include endian.h at the beginning.

File: include/nuttx/compiler.h (hunk 15 above)
Comment (xiaoxiang781216) [reply include/nuttx/compiler.h:69]: it's better to create a new patch and PR instead

File: include/nuttx/compiler.h [hunk 20]
```diff
@@ -69,10 +77,30 @@
 #  undef CONFIG_HAVE_CXX14
 #endif
 
+#if defined(__cplusplus) && __cplusplus >= 201703L
```
Comment (xiaoxiang781216) [inline include/nuttx/compiler.h:80]: move to other patch

File: libs/libc/string/lib_ffs.c [hunk 21]
```diff
@@ -48,14 +48,15 @@
  *   0, then ffs() will return 0.
  *
  ****************************************************************************/
-
```
Comment (xiaoxiang781216) [inline libs/libc/string/lib_ffs.c:51]: revert the change

File: libs/libc/string/lib_ffsl.c [hunk 22]
```diff
@@ -48,14 +48,15 @@
  *   0, then ffsl() will return 0.
  *
  ****************************************************************************/
-
```
Comment (xiaoxiang781216) [inline libs/libc/string/lib_ffsl.c:51]: revert

File: include/nuttx/compiler.h (hunk 3 above)
Comment (pkarashchenko) [reply include/nuttx/compiler.h]: Yes. For Clang you can even redefine `__GNUC__` to a desired value, so with compiler option you can reach the desired functionality.

**Comment by xiaoxiang781216** [comment]
@nimish why close this PR? will you create a new one?

**Comment by nimish** [comment]
bizarre. no idea why this closed

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by xiaoxiang781216** [comment]
@nimish the change looks good now, but a minor style issue need be addressed before merging:
https://github.com/apache/incubator-nuttx/runs/7067463510?check_suite_focus=true

File: include/endian.h [hunk 23]
```diff
@@ -51,23 +51,40 @@
...
+#  else
+#    define __swap_uint64(n) \
+        (uint64_t)(((((uint64_t)(n)) & 0x00000000000000ffULL) << 56) | \
+                ((((uint64_t)(n)) & 0x000000000000ff00ULL) << 40) | \
+                ((((uint64_t)(n)) & 0x0000000000ff0000ULL) << 24) | \
+                ((((uint64_t)(n)) & 0x00000000ff000000ULL) <<  8) | \
+                ((((uint64_t)(n)) & 0x000000ff00000000ULL) >>  8) | \
+                ((((uint64_t)(n)) & 0x0000ff0000000000ULL) >> 24) | \
+                ((((uint64_t)(n)) & 0x00ff000000000000ULL) >> 40) | \
+                ((((uint64_t)(n)) & 0xff00000000000000ULL) >> 56))
```
Comment (pkarashchenko) [inline include/endian.h:84]:
```suggestion
        (uint64_t)(((((uint64_t)(n)) & 0x00000000000000ffULL) << 56) | \
                   ((((uint64_t)(n)) & 0x000000000000ff00ULL) << 40) | \
                   ((((uint64_t)(n)) & 0x0000000000ff0000ULL) << 24) | \
                   ((((uint64_t)(n)) & 0x00000000ff000000ULL) <<  8) | \
                   ((((uint64_t)(n)) & 0x000000ff00000000ULL) >>  8) | \
                   ((((uint64_t)(n)) & 0x0000ff0000000000ULL) >> 24) | \
                   ((((uint64_t)(n)) & 0x00ff000000000000ULL) >> 40) | \
                   ((((uint64_t)(n)) & 0xff00000000000000ULL) >> 56))
```

**Review by pkarashchenko — COMMENTED** [review]
looks good in general

**Comment by pkarashchenko** [comment]
@xiaoxiang781216 looking into CI I see some warnings like
```
warning: (NET_TCP_NO_STACK && NET_USRSOCK_TCP) selects NET_TCP which has unmet direct dependencies (NET && SCHED_WORKQUEUE)
```
I think that is related to one of latest reworks in TCP TX poll timer. Could you please investigate it?

**Comment by xiaoxiang781216** [comment]
@nimish please squash the second patch into the first.

**Comment by xiaoxiang781216** [comment]
> @xiaoxiang781216 looking into CI I see some warnings like
> 
> ```
> warning: (NET_TCP_NO_STACK && NET_USRSOCK_TCP) selects NET_TCP which has unmet direct dependencies (NET && SCHED_WORKQUEUE)
> ```
> 
> I think that is related to one of latest reworks in TCP TX poll timer. Could you please investigate it?

I will take look tomorrow.

**Comment by xiaoxiang781216** [comment]
> @nimish please squash the second patch into the first.

and the style warning:
```
../nuttx/tools/checkpatch.sh -m -g 4c2b7f3ac9052d50aa1c8fa20e643ce4bbf639c6..HEAD
Error: /home/runner/work/incubator-nuttx/incubator-nuttx/nuttx/include/nuttx/compiler.h:120:1: error: Too many blank lines
```

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by xiaoxiang781216** [comment]
@nimish The inclusion of endian/Make.defs need be removed too:
```
Makefile:30: endian/Make.defs: No such file or directory
```

**Review by xiaoxiang781216 — APPROVED** [review]
